Fix internal vrp import

braktar opened the pull request.

Commit:
Allow synchrone if threshold is zero

Commit:
Provide rest timewindows compatible with vehicle

Commit:
Ortools can only deal with exactly one timewindow per rest

senhalil committed:
Fix typo introduced during API module split

senhalil committed:
Dont refuse ortools due to initial soln + schedule

senhalil committed:
Fix: periodic relation expansion

min/max_day_lapse relation can be defined based on the first visit
if and only if the lapses are equal. Otherwise, they need to be defined
via sequential visites -- which can be done via one single relation,
instead of defining each pair seperately.

senhalil committed:
Fix: schedule global day compatibility filter

senhalil committed:
Fix vehicles_indices skill compatibility condition

senhalil committed:
Fix overwritten existing periodic initial routes

senhalil committed:
Fix uninitialized variable warnings

senhalil committed:
Qual: Simplify periodic serv, vehic & relation gen

senhalil committed:
API: Do not overwrite defaults if not needed

senhalil committed:
Fix: periodic heuristic expands existing routes

senhalil committed:
Point object already has matrix_index

senhalil committed:
Fix: unfeasible detection due to unavailable days

senhalil committed:
Functions manage their extreme cases themselves

senhalil committed:
Add: keep a list of service-vehicle compatibility

Commit:
Replace marshal load/dump by a new vrp creation

fab-girard committed:
Use of mapotempo rubocop for uniform coding rules

Commit:
solution cost and cost info are sum of route costs

Commit:
StopDepot for depots

Commit:
zip cluster cannot handle activities

Commit:
Rework cluster_threshold conditions & tests

Commit:
Fix missing rest simplication_active

Commit:
maximum lateness is only a problem if we consider lateness

Commit:
Remove assert_vehicles_no_zero_duration

Commit:
Let synchronous resolution return a geojson

Commit:
Remove tmp folder

senhalil committed:
Move adjust_independent_duration t SplitClustering

senhalil committed:
partitions replaces partitn_method/metric/centroid

Commit:
Empty solution for periodic generates duplicate id error

Commit:
Fix build_vehicle id duplicate

Commit:
Fix count used routes

Commit:
Allow to exclude fields from as_json

Commit:
as_json transform skills into string

Commit:
Add cost_info exclusion field and == operator

senhalil committed:
Split duration among partitions correctly

Commit:
Fix vrp creation through as_json

Commit:
sub problems with no services

Commit:
Migrate wiki to docs/

Commit:
Migrate rubocop tests to Actions

Commit:
Setup duration is not a travel time

Commit:
Exclude already known warnings

Co-authored-by: Halil Sen <halil.sen@mapotempo.com>

Commit:
remove redundant cop disable

senhalil committed:
Make debugging easier with as_json output

senhalil commented on models/concerns/as_json.rb:
```diff
@@ -29,6 +29,17 @@ def empty_json(options = nil)
   end
 end
 
+module ServiceAsJson
+  extend ActiveSupport::Concern
+
+  def as_json(options = nil)
+    service = super
+
+    service.delete('activities') if activities.empty?
+    service
+  end
+end
+
```
Why don't we do this systematically for all dumped objects? Do we ever want to dump empty arrays? If we do it `ServiceAsJson` way, we go back to square-zero in the maintainability and automation sense.

braktar replied:
Dumping an empty array is not really an issue in the general case. Empty arrays are generated through the has_many active_hash method.

The special cases are for service activities and vehicle sequence_timewindows because these fields are mutually exclusive with their singular version...

braktar replied:
See the comments both in models/service and models/vehicle. These classes should be remove as soon as we merge singular and plural fields

senhalil reviewed (commented):
If renaming `unassigned` touches 30+ files, lets do it in the original commit of the "Models::Solution" PR because this is needlessly cluttering. Don't you agree? I just checked the other commits, it looks like cherry-picking this commit to the other mode-solution-2 branch would create no-or-very-little conflict. That way model-solution-2 would be more self contained.

As a side note: Are we sure of `unassigned_stops` because if a service is not served, does it really a "stop"? On optimizer-api everything is based on "services", so I find `unassigned_services` a better fit (but if you are worried about "shipments") then we can do `unassigned_missions` (we eventually want to rename the service/shipment into missions to become more general and inclusive anyways and we already have "mission_ids" in routes; so it wouldn't be to out of place I believe) what do you think?

But that being said, at the end of the day this is internal code, right? So my preference would be `unassigned_services` (we only have services internally), `unassigneds` (which is ungrammatical but shorter), `unassigned_missions`.

Commit:
Integrate comments from PR 335

Commit:
Some rephrasing

Commit:
Once OR-Tools has finished pids should be reset

fab-girard committed:
Merge pull request #321 from braktar/model-solution-2

Model solution 2

Commit:
Edit changelog

senhalil committed:
Fix ids function naming in base.rb

Commit:
Edit changelog

Commit:
Edit changelog

senhalil committed:
Merge pull request #361 from senhalil/dev

- Base debugging output to as_json to prevent full obj prints
- Fix ids function naming in base.rb

senhalil committed:
Exclude local config initializer

senhalil committed:
Fix test warnings

senhalil committed:
Caching is on by default and managed via .cache

senhalil committed:
Merge pull request #331 from fab-girard/quality

Start using mapotempo_rubocop gem

senhalil committed:
Merge pull request 360 from braktar/fix-rest-simplification

Fix rest simplification

senhalil committed:
Update config/locales/en.yml

senhalil committed:
Merge pull request 348 from braktar/trad-compl

Finish api-front translation

dependabot[bot] committed:
Bump puma from 5.5.2 to 5.6.2

Bumps [puma](https://github.com/puma/puma) from 5.5.2 to 5.6.2.
- [Release notes](https://github.com/puma/puma/releases)
- [Changelog](https://github.com/puma/puma/blob/master/History.md)
- [Commits](https://github.com/puma/puma/compare/v5.5.2...v5.6.2)

---
updated-dependencies:
- dependency-name: puma
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

senhalil committed:
Merge pull request 347 from braktar/reset-pids

Once OR-Tools has finished pids should be reset

Commit:
Do not provide default values

senhalil committed:
Merge pull request 357 from braktar/remove-assert-zero-duration

Remove assert_vehicles_no_zero_duration

senhalil committed:
Merge pull request 355 from braktar/remove-tmp-folder

Remove tmp folder

Commit:
Avoid to_json/Json.parse

Commit:
Add docs link to readme

senhalil committed:
Merge pull request 354 from braktar/empty-solution-periodic

Empty solution for periodic generates duplicate id error

senhalil committed:
Merge pull request 351 from braktar/migrate-wiki

Migrate wiki to docs

senhalil committed:
Merge pull request 356 from braktar/post-get-geojsons

Let synchronous resolution return a geojson

senhalil committed:
Merge pull request 358 from braktar/stop-depot

StopDepot for depots

senhalil committed:
Merge pull request 353 from braktar/marshal-dump-repetitions

Marshal dump repetitions

senhalil committed:
Merge pull request 359 from braktar/remove_cost_field

Remove cost field

senhalil committed:
Merge pull request 336 from senhalil/feat/fix_#363_partition_resolution_time

Fix #363 partition resolution time

senhalil committed:
Bump optimizer-ortools to v1.12.0

senhalil committed:
Update CHANGELOG.md

senhalil committed:
Test repeat a failing performance test

senhalil committed:
Extract activity.duration on vehicle calculation

senhalil committed:
Merge pull request 318 from senhalil/feat/general_fixes_and_vehicle_compatibility_check_improvement

[optim-ortools] General fixes and vehicle compatibility check improvement

senhalil committed:
Merge pull request 346 from Mapotempo/dependabot/bundler/puma-5.6.2

Bump puma from 5.5.2 to 5.6.2

Commit:
Extend multiple vehicle duplicate logic of the same day_index

Commit:
move to_sym to model create and as_json_fields

senhalil committed:
Merge pull request 352 from braktar/cost-info-operator

Cost info operator

braktar committed:
travel_time_correction should be integer

Co-authored-by: Halil Sen <halil.sen@mapotempo.com>

Commit:
Merge pull request #350 from braktar/travel-time-setup-duration

Setup duration is not a travel time

Commit:
Avoid Filtering of partial sub_problems

senhalil committed:
Fix: division by 0

senhalil committed:
Negative value provided by router is not a warning

senhalil committed:
No need to have a warning if it is normal

senhalil committed:
Merge pull request 365 from senhalil/dev

Fix: division by 0
Negative value provided by router is not a warning
No need to have a warning if it is normal

Commit:
Setup duration related to matrix index

Pierre-Graber committed:
Ignore local IDE settings

senhalil committed:
Merge pull request 368 from Pierre-Graber/dev

Ignore local IDE settings

senhalil committed:
Capacity limit cannot be negative

senhalil committed:
polyline generation posibility is default on local

senhalil committed:
Modify empty/fill to fix current_load calculation

Commit:
Define field type

senhalil committed:
Merge pull request 370 from senhalil/dev

 Modiy empty/fill logic to fix current_load calculation

senhalil committed:
Merge pull request 369 from braktar/setup-matrix-index

Setup duration related to matrix index

senhalil committed:
Merge pull request 363 from braktar/schedule-vehicle-name

Schedule vehicle name

Commit:
Return typed fields with as_json

Commit:
Add validate endpoint

Commit:
Linked_ids and linked_service_ids are currently most of the time duplicate values

Commit:
Edit changelog

senhalil committed:
Either activity or activities is required

senhalil committed:
activity or activities cannot be empty

senhalil committed:
Merge pull request 377 from senhalil/dev

Either activity or activities is required

senhalil committed:
Apply suggestions from code review

senhalil committed:
Merge pull request 217 from fonsecadeline/deal_with_threshold_zero

Allow synchrone if threshold is zero

senhalil committed:
Add ripper-tags for offline code analysis

senhalil committed:
Forward accumulate skills of linked services

This is to make sure that split_independent_vrp respects
linking_relations and infeasible services (those with incompatible
skills) will be eliminated by detect_unfeasible (if there is no vehicle
than can service linked services).

senhalil committed:
Update CHANGELOG.md

senhalil committed:
Qual: fix cop ordering

senhalil committed:
Qual: disable Style/GuardClause

senhalil committed:
Update .rubocop.yml

senhalil committed:
Merge pull request 379 from senhalil/dev

Closes https://gitlab.com/mapotempo/optimizer-api/-/issues/912

senhalil committed:
Update CHANGELOG.md

senhalil committed:
Merge branch 'dev' into provided_consistent_rest_tws

senhalil committed:
Merge pull request 273 from fonsecadeline/provided_consistent_rest_tws

Provide rest timewindows compatible with vehicle

Commit:
Correctly collect sub_vrp rests

Commit:
Routes and cluster_threshold

Commit:
Remove skills create logic and extend convert logic
